File: backend/src/main/java/pers/eastwind/billmanager/attach/controller/AttachmentController.java
package pers.eastwind.billmanager.attach.controller;

import org.springframework.core.io.Resource;
import org.springframework.http.*;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import pers.eastwind.billmanager.attach.model.AttachmentDTO;
import pers.eastwind.billmanager.attach.service.AttachmentService;
import pers.eastwind.billmanager.common.exception.BizException;

import java.util.ArrayList;
import java.util.List;

/**
 * 附件控制器
 */
@RequestMapping("/api/attachment")
@RestController
public class AttachmentController {
    private final AttachmentService attachmentService;

    public AttachmentController(AttachmentService attachmentService) {
        this.attachmentService = attachmentService;
    }

    /**
     * 下载文件
     *
     * @param attachment 文件 DTO
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'USER', 'FINANCE')")
    @GetMapping(value = "/")
    public ResponseEntity<Resource> download(AttachmentDTO attachment) {
        if (attachment == null) {
            throw new BizException("不能为空");
        }
        Resource resource = attachmentService.getResource(attachment);
        HttpHeaders headers = new HttpHeaders();
        headers.setContentDisposition(ContentDisposition.attachment().name(attachment.getName()).build());
        headers.setContentType(MediaType.APPLICATION_OCTET_STREAM);
        return new ResponseEntity<>(resource, headers, HttpStatus.OK);
    }

    /**
     * 上传文件至临时路径
     *
     * @param files 文件
     * @return 文件信息
     */
    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
    @PostMapping("/temp")
    public List<AttachmentDTO> uploadTemp(@RequestParam List<MultipartFile> files) {
        List<Resource> resources = new ArrayList<>();
        for (MultipartFile file : files) {
            resources.add(file.getResource());
        }
        return attachmentService.uploadTemps(resources);
    }
}

File: backend-go/internal/module/attach/controller/attachment_controller.go
package controller

import (
	"backend-go/internal/module/attach/model"
	"backend-go/internal/module/attach/service"
	"backend-go/internal/pkg/errs"
	"backend-go/internal/pkg/result"

	"github.com/gofiber/fiber/v3"
)

type AttachmentController struct {
	attachService *service.AttachmentService
}

func NewAttachmentController(attachService *service.AttachmentService) *AttachmentController {
	return &AttachmentController{
		attachService: attachService,
	}
}

func (c *AttachmentController) UploadTemp(ctx fiber.Ctx) error {
	form, err := ctx.MultipartForm()
	if err != nil {
		return err
	}
	files := form.File["files"]
	if len(files) == 0 {
		return errs.NewBizError("没有文件上传")
	}
	temps, err := c.attachService.UploadTemps(files)
	if err != nil {
		return err
	}

	result.SetResult(ctx, temps)
	return nil
}

func (c *AttachmentController) Download(ctx fiber.Ctx) error {
	var dto model.AttachmentDTO
	if err := ctx.Bind().Query(&dto); err != nil {
		return err
	}
	name, path, err := c.attachService.GetResource(ctx, &dto)
	if err != nil {
		return err
	}
	return ctx.Download(path, name)
}
